Update Governance documentation (#7431)
elastic/eui: GOVERNANCE.md

# EUI’s Design System Governance

> This Governance describes the processes we follow within Elastic when working with the EUI Design System.

Reviewer: ```suggestion
This governance document describes the processes we follow within Elastic when working with the EUI design system.
```

## Prelude and principles

The following are some quotes and principles that guide the way we think about EUI.

> You can have a comprehensive design system [...] but if a design system user can’t get done what they’re trying to get done, the whole system risks obsolescence.
>
> Teams get creative and will find ways of getting things done, which may involve hacking styles, creating a slew of one-off components, or abandoning the design system altogether.
>
> This is why it is so incredibly important [...] to establish a crystal-clear governance process that helps users understand what to do when:
>
> - They can’t find a component that does what they need.
> - A design system component gets them 90% of the way there, but not 100%.
> - They have questions or concerns about the design system.

\- [Brad Frost](https://bradfrost.com/blog/post/a-design-system-governance-process/)

> > Design Systems should actively discourage deviations from the system, otherwise all the benefits of the system are lost.

Reviewer: ```suggestion
> Design systems should actively discourage deviations from the system, otherwise all the benefits of the system are lost.
```

> \- Someone, somewhere, probably

Reviewer: lmao, I love this so much ✨

> - Enablement over enforcement: EUI is here to help you move faster, not to put up walls or slow you down. It works best when we collaborate early and often. "When in doubt, react out!"

Reviewer: ```suggestion
- Enablement over enforcement: EUI is here to help you move faster, not to put up walls or slow you down. It works best when we collaborate early and often. "When in doubt, reach out!"
```
I think this is what you meant? unless this is a genius pun on React, in which case, carry on

2. We will establish a consistent product UX, regardless of solution or feature. This has an overwhlemingly positive effect on user perception of our products.
3. We will have less redundant code.
4. It's much easier to make sweeping product UX changes when we're using EUI consistently.

Things we want to avoid:
1. **Applying custom CSS to work around EUI** - These changes are brittle and prone to break in the future. They create UX inconsistencies.
2. **Creating new UI patterns and elements** - New UI elements and patterns breaks user consistency, creates more work for engineers, and opens the door for regressions and quality issues. EUI is well tested and built for quality and accessibility by default. EUI may not work for every scenario but please make an effort to use what we have when possible.

### 2. When in doubt, reach out!

The EUI team is here to help you work within the EUI design system. If you're not sure if we have a capability, what component to use, or how to achieve a certain result, please reach out! A short conversation with us could save you a lot of work downstream.

You can interact with us by:
- Subbmitting a Github issue

> - Contacting us at #eui on Slack (Internal to Elastic)

Reviewer: ```suggestion
- Contacting us at #eui on Slack (internal to Elastic)
```

Ideally, we can help you find a way to work within what EUI provides. If not, proceed to step 3 or 4!

### 3. When we have something that works for you, but would require customization for your use case

> For example -- you've found EUIBadge. It's perfect. Except for one thing ... it only provides solid colors, and you really feel that for your use case, it should be not be solid shaded and just have a colored border.

Reviewer: ```suggestion
For example -- you've found `EuiBadge`. It's perfect. Except for one thing ... it only provides solid colors, and you really feel that for your use case, it should be not be solid shaded and just have a colored border.
```

Firstly, congratuations, at this point you've followed step 1, you're using EUI.

Secondly, you've likely reached out to the EUI team to start a conversation about this when you realized it didn't quite work for you, well done!

In this case, the first thing you probably heard from the EUI team is that we *really* want you to consider using a different approach. We really feel like you should choose a different UX treatment, one that that does not require customizing EUIBadge.

However, after some discussion with our team, you've convinced us that this is truly the only thing that will work here. At this point, we will:

1. Provide you a workaround with custom CSS for a workaround.
2. Track this as a [one-off](#one-offs) customization.

> If you feel strongly that this customization should be part of EUI and aligns with our guidance for [What we do accept](#what-we-do-and-dont-accept), you should create a [Feature Request](https://github.com/elastic/eui/issues/new?assignees=&labels=feature+request&projects=&template=feature_request.md&title=) for it.

Reviewer: ```suggestion
If you feel strongly that this customization should be part of EUI and aligns with our guidance for [What we do accept](#what-we-do-and-dont-accept), you should create a [feature request](https://github.com/elastic/eui/issues/new?assignees=&labels=feature+request&projects=&template=feature_request.md&title=) for it.
```

> The EUI team will consider the Feature Request and assign it a priority and work it as we have time. If and when we do, we will be sure to circle back to your "one-off" to make sure it is replaced with this new feature.

Reviewer: ```suggestion
The EUI team will consider the feature request and assign it a priority and work it as we have time. If and when we do, we will be sure to circle back to your "one-off" to make sure it is replaced with this new feature.
```

Not sure I have a good suggestion for this, but the second sentence implies that **we** are the ones that will be replacing the one off with the new feature. Is that the intention, or should we leave some room in there to imply that other teams should also own or be responsible for replacing their own one-offs?

Author: It's a good point. I could imagine that sometimes we may just do it, other times we may ask a team to do it.

We probably need to think more about how we track and clean up one-offs.

Reviewer: My personal preference would be to word it as if the default is to ask teams to do it. That way the bar is real low and if we do end up being able to be the ones to help out, we get extra credit :trollface: 

```suggestion
The EUI team will consider the feature request and assign it a priority and work it as we have time. If and when we do, we will be sure to circle back to your "one-off" to request that it is replaced with the new feature.
```

- Features should have clear guidance around *when* they should be used.
- Bug fixes :D

**What we don’t accept**
- Features specific to a single use case (see [One-offs](#one-offs), below)
- Features that carry an excessive long-term maintenance burden, relative to their usage.
- Features that are inaccessible.

**Open source**
Elastic is an open-source repository, but not completely open-contribution. While we appreciate and welcome all your ideas and contributions, we cannot always guarantee we’ll accept them. Please see our [CONTRIBUTING.md](./wiki/contributing-to-eui/README.md) for more.

## One-offs

One-off components are a reality in mature design systems. If existing components do not meet your needs, it is acceptable to create a one-off components. One-offs must meet the following criteria to be considered for inclusion in EUI when its usage grows.
- The component must match the look and feel of EUI components
- The component follows EUI best practices, especially UX and accessibility
- The component has undergone manual accessibility testing
- The component has or is included in automated accessibility tests

> - Works with Dark Mode

Reviewer: ```suggestion
- Works in dark mode
```
